MaxText/maxtext_utils.py

```diff
@@ -0,0 +1,417 @@
+    if not state:
+      init_train_state_partial = functools.partial(init_train_state, model, tx, config)
+      state = jax.jit(
+          init_train_state_partial,
+          in_shardings=None,
+          out_shardings=state_mesh_shardings
+      )(rng)
+      if raw_params: # If we loaded a partial state, we need to merge it.
+        state = state.replace(params = raw_params)
+    raw_params = None
+
+  state = unbox_logicallypartioned_trainstate(state)
+  return state, state_mesh_annotations
+
+def create_device_mesh(config, devices=jax.devices(), logging=True):
+  """Creates a device mesh with each slice in its own data parallel group. If there is only one slice, uses two replicas """
+  num_devices = len(devices)
+  try:
+    num_slices = 1 + max([d.slice_index for d in devices])
+  except:
+    num_slices = 1
+  num_devices_per_slice = num_devices//num_slices
+  max_logging.log(f"Devices: {devices} (num_devices: {num_devices})")
+  assert len(devices) > 1, "You must have at least two devices"
+
+  multi_slice_env = num_slices > 1
+
+  dcn_parallelism = [config.dcn_data_parallelism, config.dcn_fsdp_parallelism, config.dcn_tensor_parallelism]
+  ici_parallelism = [config.ici_data_parallelism, config.ici_fsdp_parallelism, config.ici_tensor_parallelism]
+
+  # Find possible unspecified parallelisms
+  dcn_parallelism = fill_unspecified_mesh_axes(dcn_parallelism, num_slices, 'DCN')
+  ici_parallelism = fill_unspecified_mesh_axes(ici_parallelism, num_devices_per_slice, 'ICI')
+
+  if multi_slice_env:
+    mesh = mesh_utils.create_hybrid_device_mesh(ici_parallelism, dcn_parallelism, devices)
+  else:
+    mesh = mesh_utils.create_device_mesh(ici_parallelism, devices)
+
+  if logging:
+    max_logging.log(f"Decided on mesh: {mesh}")
+
+  return mesh
+
+def fill_unspecified_mesh_axes(parallelism_vals, target_product, parallelism_type):
+  """Evaluates unspecified DCN/ICI parallelism values"""
+  if -1 in parallelism_vals:
+    assert parallelism_vals.count(-1) == 1, f"Found unspecified values (-1) for more than one {parallelism_type}\
+      parallelism axis. At most one axis can be unspecified."
+
+    determined_val = target_product/np.product(parallelism_vals)*-1
+
+    assert determined_val >= 1 and determined_val.is_integer, f"Unspecified value unable to be determined with the given\
+      {parallelism_type} parallelism values"
+
+    parallelism_vals[parallelism_vals.index(-1)] = int(determined_val)
+
+  target_type = "slices" if parallelism_type == 'DCN' else "devices per slice"
+
+  assert np.product(parallelism_vals) == target_product, f"Number of {target_type} {target_product} does not match\
+    the product of the {parallelism_type} parallelism {np.product(parallelism_vals)}"
+
+  return parallelism_vals
+
+# Learning Rate Schedule
+# -----------------------------------------------------------------------------
+
+def create_learning_rate_schedule(config):
```

Comment:
Theoretically different models have different optimal learning rate schedules, but maybe its common to share across